### DISCUSSION TITLE
Was .NET5 and earlier dropped for v2.5.0?

### DISCUSSION THREAD

**Discussion by egil** [xunit/xunit#2771] [Question] [answered]
After upgrading to v2.5.0 all my tests that target net5 or netcoreapp3.1 are no longer being run and discovered in Visual Studio. Moving back to v2.4.2 reverts the change.

Did something in xUnit change or is there another reason for this behavior?

ps. could not find a changelog that describes the changes.

**Comment by bradwilson** [comment, marked as answer]
Yes. From [the README file](https://github.com/xunit/visualstudio.xunit/blob/main/README.md):

> This project contains the Visual Studio runner for xUnit.net. It supports the built-in Test Explorer feature in Visual Studio 2022 and later. It supports .NET 6+ and .NET Framework 4.6.2+.

It was dropped because .NET 5 is no longer supported by Microsoft (it has been out of support since May 2022). During the 2.5.0 release cycle, the minimum bar was moved from .NET Core 3.1 (the previous LTS) to .NET 6 (the current LTS).

Support list: https://dotnet.microsoft.com/en-us/platform/support/policy

**Reply by bradwilson** [reply]
If you're stuck on an unsupported version of .NET, you can continue to use the 2.4.5 of `xunit.runner.visualstudio`. The supported versions for the core framework itself are still _very_ old, so you can move the core framework forward if you wish (mixing 2.5.x core framework with 2.4.x VS runner does work, as our runners have both backward compatibility, as well as forward compatibility within the same major version, so 2.4.5 of the VS runner should work against any 1.x or 2.x of the core framework).

**Reply by bradwilson** [reply]
It's also worth mentioning that the prior version of the VS runner being dependent on .NET Core 3.1 was reported as a bug, reasonably so, because that required people to install an unsupported version of .NET Core on their machine even when they didn't plan to use it. That's one of the reasons it was moved forward. https://github.com/xunit/visualstudio.xunit/issues/360

**Comment by egil** [comment]
> It was dropped because .NET 5 is no longer supported by Microsoft (it has been out of support since May 2022). During the 2.5.0 release cycle, the minimum bar was moved from .NET Core 3.1 (the previous LTS) to .NET 6 (the current LTS).

Thats fair. In bUnit however, we have not yet dropped support for .NET5 and below because we have not moved to a new major version of the library, and try our best to follow semver conventions. I will just keep the runner at 2.4.2 and move the rest of the packages to 2.5.0.

Thanks!

**Comment by egil** [comment]
Hmm it does not seem to always work with vs.runner at v1.4.2 and the other libraries at v2.5.0. I am seeing the following error:

```
An exception occurred while invoking executor 'executor://xunit/VsTestRunner2/netcoreapp': Method 'LogRaw' in type 'Xunit.Runner.VisualStudio.VisualStudioRunnerLogger' from assembly 'xunit.runner.visualstudio.dotnetcore.testadapter, Version=2.4.0.0, Culture=neutral, PublicKeyToken=8d05b1bb7a6fdb6c' does not have an implementation.
Stack trace:
   at Xunit.Runner.VisualStudio.VsTestRunner.RunTests(IRunContext runContext, IFrameworkHandle frameworkHandle, LoggerHelper logger, TestPlatformContext testPlatformContext, RunSettings runSettings, Func`1 getRunInfos)
   at Xunit.Runner.VisualStudio.VsTestRunner.Microsoft.VisualStudio.TestPlatform.ObjectModel.Adapter.ITestExecutor.RunTests(IEnumerable`1 sources, IRunContext runContext, IFrameworkHandle frameworkHandle) in /_/src/xunit.runner.visualstudio/VsTestRunner.cs:line 170
   at Microsoft.VisualStudio.TestPlatform.Common.ExtensionDecorators.SerialTestRunDecorator.RunTests(IEnumerable`1 sources, IRunContext runContext, IFrameworkHandle frameworkHandle) in /_/src/Microsoft.TestPlatform.Common/ExtensionDecorators/SerialTestRunDecorator.cs:line 57
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Execution.RunTestsWithSources.InvokeExecutor(LazyExtension`2 executor, Tuple`2 executorUriExtensionTuple, RunContext runContext, IFrameworkHandle frameworkHandle) in /_/src/Microsoft.TestPlatform.CrossPlatEngine/Execution/RunTestsWithSources.cs:line 115
   at Microsoft.VisualStudio.TestPlatform.CrossPlatEngine.Execution.BaseRunTests.RunTestInternalWithExecutors(IEnumerable`1 executorUriExtensionMap, Int64 totalTests) in /_/src/Microsoft.TestPlatform.CrossPlatEngine/Execution/BaseRunTests.cs:line 466
```

**Reply by bradwilson** [reply]
What about if you use version 2.4.5? That was the latest prior to the 2.5.x move.

**Reply by egil** [reply]
I see the same error with 2.4.5 of the vs runner and 2.5.0 of the rest of the packages.

**Reply by bradwilson** [reply]
I'm not able to reproduce that here.

.csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <LangVersion>10.0</LangVersion>
    <Nullable>enable</Nullable>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.7.2" />
    <PackageReference Include="xunit" Version="2.5.0" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.5">
      <IncludeAssets>runtime;build;native;contentfiles;analyzers;buildtransitive</IncludeAssets>
      <PrivateAssets>all</PrivateAssets>
    </PackageReference>
  </ItemGroup>

</Project>
```

.cs:

```csharp
using Xunit;

public class UnitTest
{
    [Fact]
    public void Passing()
    {
        Assert.True(true);
    }

    [Fact]
    public void Failing()
    {
        Assert.True(false);
    }
}
```

Both `dotnet test` and Test Explorer in VS 2022 are working as expected:

![image](https://github.com/xunit/xunit/assets/16944/a9b71d4c-5a28-414d-b884-a0447567a7f8)

![image](https://github.com/xunit/xunit/assets/16944/56f69d2c-2ddc-4545-9ff9-be9f6116289a)

**Reply by bradwilson** [reply]
Oops, I forgot, .NET 5.0. One second while I install that and retry...

**Reply by bradwilson** [reply]
Same results:

![image](https://github.com/xunit/xunit/assets/16944/52fe781d-0e32-4f49-9a68-fb3af37923a7)

![image](https://github.com/xunit/xunit/assets/16944/67f50539-cd7d-46c8-8368-f474552a49c0)

**Reply by egil** [reply]
I'll create a branch in our repo and send you a link so you can investigate.

**Reply by egil** [reply]
I have a branch/PR now that shows the problem. This also happens when running with `dotnet test` locally on my Windows machine.

- PR/branch: https://github.com/bUnit-dev/bUnit/pull/1199
- GitHub actions run error: https://github.com/bUnit-dev/bUnit/actions/runs/6083013652/job/16502019648?pr=1199#step:7:144

**Reply by bradwilson** [reply]
I see you're linking against `xunit.runner.utility` so there's likely a conflict there, since `xunit.runner.visualstudio` also includes that and copies it into your output folder: https://github.com/xunit/visualstudio.xunit/blob/26ed962977078c0328351a0e3e0c4313ee672bd0/src/xunit.runner.visualstudio/build/xunit.runner.visualstudio.dotnetcore.props#L14-L18 That would explain the binary compatibility issue.

So you're probably stuck on `xunit` 2.4.2 as a result until you move away from .NET 5. I'll add a bug to ILMerge `xunit.runner.utility`, though that won't help you since that'll be for newer builds (which don't support .NET 5).

**Reply by bradwilson** [reply]
https://github.com/xunit/visualstudio.xunit/issues/383

**Reply by egil** [reply]
Thanks. Appreciate the help.

**Reply by bradwilson** [reply]
@egil So, bad news.

The change I did in 383 broke 3rd party reporters. There's no way for me to reconcile those changes and have third party reporters still work. Here is my description of the problem over there: https://github.com/xunit/visualstudio.xunit/issues/385#issuecomment-1741900894

I think the design of the system today is broken because of this need to share `xunit.runner.utility` for both runners and runner extenders, because we've made no effort to ensure that we don't have breaking changes between versions of `xunit.runner.utility`; specifically, we would need a frozen assembly like `xunit.abstractions` to define the extensibility points here, but the time to do that was about 8½ years ago. The need is the same (frozen contracts) but for different reasons (`xunit.abstractions` exists as it does today because of the frozen contracts that cross the AppDomain boundaries between the runner and the test execution engine; the hypothetical interface-definition library for third party runner extensibility would need to be frozen so that its contract never changed).

In theory, the problem of how to test a runner from a runner in the past hasn't been an issue, because with AppDomain isolation you could have separate versions of `xunit.runner.utility`, but that changed with VSTest and .NET Core. No more AppDomains means no more isolation, and since .NET Core unit test projects in v2 are not separate processes, you have this assembly collision. Worse, the compiler can't help you here, because you don't _link_ against `xunit.runner.visualstudio`; instead, we copy those binaries into your output folder because there's no other way for VSTest to work outside of Visual Studio (inside of Visual Studio, it relies on its ability to traverse your package references via the object model that VS provides, but this capability does not exist from the command line, and definitely not outside of Windows).

So there's no way for me to solve this problem. Unfortunately, I have to re-break you to un-break others. You will be suffering from the same problem I describe in the other issue of conflicting versions of `xunit.runner.utility` maybe not being compatible. It may work, and it may not; you're basically running a gamble without hard-sticking your `xunit.runner.utility` dependency to be the exact same version as what's in the version of `xunit.runner.visualstudio` that you choose to use. I don't know how much that might break the design of bUnit.

Depending on what your consumption of `xunit.runner.utility` means, you may be able to take the approach I did in 383 and use ILRepack to merge `xunit.runner.utility` directly into your bUnit libraries, and then the collision disappears (for the same reason as we accomplished in 383, but on the reverse end of the relationship). This would work if the reason you're consuming `xunit.runner.utility` is to _**be a runner**_ and not to _**plug into our runners**_.

I'm about to commit the revert of 383, which may be useful for you if you wish to go down this path.

Apologies, but there's really no good path here. Someone's gotta break, and undoing the work from 383 at least gives all of us the possibility to start working again, albeit with more work on your end.

**Reply by egil** [reply]
@bradwilson, appreciate the effort. Please don't make any changes on our behalf. The part of our codebase that uses xunit.runner.utility is a legacy package we are keeping around for completeness sake, and it could just as well be deprecated and taken out of our build step completely.
